File: Python/10_First non-repeating character in a stream.py
# First non-repeating character in a stream
# https://practice.geeksforgeeks.org/problems/first-non-repeating-character-in-a-stream1216/1

from collections import Counter

class Solution:
    def FirstNonRepeating(self, A):
        # Code here
        
        ans=""
        d={}
        main=[]
        for i in A:
            d[i]=d.get(i,0)+1
            if i not in main and d[i]==1:
                main.append(i)
            if i in main and d[i]>1:
                main.remove(i)
            if ans=="" or ans[-1]==i or ans[-1]=='#':
                if main:
                    ans+=main[0]
                else:
                    ans+='#'
            else:
                ans+=ans[-1]
        return ans

#{ 
#  Driver Code Starts
#Initial Template for Python 3

if __name__ == '__main__':
    T=int(input())
    for i in range(T):
        A = input()
        ob = Solution()
        ans = ob.FirstNonRepeating(A)
        print(ans)

# } Driver Code Ends

File: C++/10_First_Non_Repeating_Character_In_A_Stream.cpp
// # First non-repeating character in a stream
// # https://practice.geeksforgeeks.org/problems/first-non-repeating-character-in-a-stream1216/1

//Approach :-
/*
    To find the first non-repeating character in a stream, we need to create a hash which would
    store the count of each character in the stream uptill now. We will also have a queue the front
    element of which always stores the first non repeating character.
    We iterate through the string given,and push the current character in our queue. Then we increment
    the count of current character by 1.Now we look at the front element of queue. If the count
    of front element is greater than 1, we pop it since it cannot be a non repeating character. We 
    continue this process until we find an element with count = 1(which would be the ans), or the queue 
    becomes empty in which case no non repeating character exists.

    Time Complexity :- O(N)
    Space Complexity :- O(N)
*/

//Code :- 
class Solution {
	public:
		string FirstNonRepeating(string A){
		    // Code here
		    int n = A.length();
		    vector<int>count(26,0);
		    queue<char>q;
		    string res;
		    for(int i=0;i<n;i++){
		        q.push(A[i]);
		        count[A[i]-'a']++;
		        while(!q.empty() && count[q.front()-'a']>1){
		            q.pop();
		        }
		        res.push_back((q.empty()?'#':q.front()));
		    }
		    
		    return res;
		}

};